rfcs/0082-lib-experimental.md

```diff
@@ -0,0 +1,129 @@
+---
+feature: lib.experimental
+start-date: 2020-12-17
+author: Francois-Rene Rideau
+co-authors: N/A
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+I propose that we create a directory lib/experimental for experimental code,
+also available as lib.experimental and pkgs.lib.experimental. Such code is
+available for limited use within select parts of nixpkgs, with the explicit
+opt-in approval of their respective maintainers, but not for widespread use
+in other parts of nixpkgs.
+
+# Motivation
+[motivation]: #motivation
+
+Some libraries we *must* use
+mandatorily unless it's impossible or at least unreasonable not to.
+Some libraries we *should* use
+unless there's a good reason not to.
+Some libraries we *may* use
+according to each contributors' preferences.
+Some libraries we *may not* use
+without an explicit endorsement by enough maintainers of the relevant packages.
```

Comment:
Should I read this as "forbidden"?

Nixpkgs doesn't only define packages. Most of the interesting use cases for `lib` functions are in `nixos/` (calling into `lib.modules`, `lib.types` and various functions to support the configuration logic). Some of the functions primarily support users' own projects, or at least that's how I think of `lib.cleanSourceWith`.

Usually when you choose to rely on someone's open source function, you accept that you may have to help improve it when the need arises. This is a mutually beneficial situation. You can make use of whatever beneficial qualities that code already has and when you need an improvement, the others will benefit as well.

When you decrease the appeal of certain functions, they will receive less use, and therefore less attention.
In my view, use of Nixpkgs is already quite pragmatic, in the sense that contributions arise from users' specific needs. So my questions is, what is the cost of keeping users away?

Reply:
I mean that, assuming my pop library gets in, and someone wants to use it while sending a patch to some recipe that uses a different extension mechanism, I expect the recipe's owner to reject it and request the patch author to use the previously used extension mechanism instead. Alternatively, if daring, he'd ask the patch author to rewrite everything using pop.